### PR TITLE
bugfixes: {segfaults, hashes, abstract unix sockets}

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: src/lxc/monitor.c [hunk 1]
```diff
@@ -153,36 +153,52 @@ int lxc_monitor_close(int fd)
...
+		ERROR("failed to create name for monitor socket");
 		return -1;
 	}
 
+	/* Note: snprintf() will \0-terminate addr->sun_path on the 106th byte
+	 * and so the abstract socket name has 105 "meaningful" characters. This
+	 * is absolutely intentional. For further info read the comment for this
+	 * function above!
+	 */
 	len = sizeof(addr->sun_path) - 1;
```
Comment (brauner) [inline src/lxc/monitor.c]: Note that we could easily guarantee that monitors created by 105-char liblxc's are attachable for 106-char liblxc's. It's the reverse case that is not easily dealt with. It's better to stick with the 105-char version.

Comment (brauner) [inline src/lxc/monitor.c]: This is obviously only a problem for lxcpaths >= sizeof(sun_path) (or sizeof(sun_path) - 3 to be pedantic. -1 for the starting \0-byte and another -1 for the terminating \0-byte that we use to be able to print out the abstract socket name in the logs, and another -1 for the \0-byte caused by incorrect checks in af_unix.c).

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

File: src/lxc/af_unix.c [hunk 2]
```diff
@@ -55,7 +55,7 @@ int lxc_abstract_unix_open(const char *path, int type, int flags)
 
 	addr.sun_family = AF_UNIX;
 
-	len = strlen(&path[1]) + 1;
+	len = strlen(&path[1]);
 	if (len >= sizeof(addr.sun_path) - 1) {
```
Comment (0x0916) [inline src/lxc/af_unix.c]: In these low-level functions, we allow the socket to be 107 length, so the check should be `if (len >= sizeof(addr.sun_path)) {`

Comment (brauner) [inline src/lxc/af_unix.c]: `sizeof(addr.sun_path)` will give you `108` so `len(bla) >= 108 - 1` is correct.

Comment (0x0916) [inline src/lxc/af_unix.c]: yes, I know, but `len(bla)` can be 107, I think it is valid socket name at this situation. The current code will return `ENAMETOOLONG` if `len = 107`.

Comment (brauner) [inline src/lxc/af_unix.c]: Yeah, at least if we don't want to enforce \0-termination. Which we probably shouldn't here.

Comment (brauner) [inline src/lxc/af_unix.c]: Thanks!

**Pull request merged**